File: software/examples/csharp/ExampleCallback.cs
using System;
using Tinkerforge;

class Example
{
	private static string HOST = "localhost";
	private static int PORT = 4223;
	private static string UID = "XXYYZZ"; // Change XXYYZZ to the UID of your Stepper Brick

	private static Random random = new Random();

	// Use position reached callback to program random movement
	static void PositionReachedCB(BrickStepper sender, int position)
	{
		int steps;

		if(random.Next(0, 2) == 0)
		{
			steps = random.Next(1000, 5001); // steps (forward)
			Console.WriteLine("Driving forward: " + steps + " steps");
		}
		else
		{
			steps = random.Next(-5000, -999); // steps (backward)
			Console.WriteLine("Driving backward: " + steps + " steps");
		}

		int vel = random.Next(200, 2001); // steps/s
		int acc = random.Next(100, 1001); // steps/s^2
		int dec = random.Next(100, 1001); // steps/s^2

		Console.WriteLine("Configuration (vel, acc, dec): (" +
		                  vel + ", " + acc + ", " + dec + ")");

		sender.SetSpeedRamping(acc, dec);
		sender.SetMaxVelocity(vel);
		sender.SetSteps(steps);
	}

	static void Main()
	{
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickStepper stepper = new BrickStepper(UID, ipcon); // Create device object

		ipcon.Connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Register position reached callback to function PositionReachedCB
		stepper.PositionReachedCallback += PositionReachedCB;

		stepper.Enable(); // Enable motor power
		stepper.SetSteps(1); // Drive one step forward to get things going

		Console.WriteLine("Press enter to exit");
		Console.ReadLine();
		stepper.Disable();
		ipcon.Disconnect();
	}
}

File: software/examples/java/ExampleCallback.java
import java.util.Random;

import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickStepper;
import com.tinkerforge.TinkerforgeException;

public class ExampleCallback {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;

	// Change XXYYZZ to the UID of your Stepper Brick
	private static final String UID = "XXYYZZ";

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		// Note: Declare stepper final, so the listener can access it
		final BrickStepper stepper = new BrickStepper(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Use position reached callback to program random movement
		stepper.addPositionReachedListener(new BrickStepper.PositionReachedListener() {
			Random random = new Random();

			public void positionReached(int position) {
				int steps = 0;
				if(random.nextInt(2) == 1) {
					steps = random.nextInt(4001) + 1000; // steps (forward)
				} else {
					steps = random.nextInt(5001) - 6000; // steps (backward)
				}

				int vel = random.nextInt(1801) + 200; // steps/s
				int acc = random.nextInt(901) + 100; // steps/s^2
				int dec = random.nextInt(901) + 100; // steps/s^2
				System.out.println("Configuration (vel, acc, dec): (" + 
				                   vel + ", " + acc + ",  " + dec + ")");

				try {
					stepper.setSpeedRamping(acc, dec);
					stepper.setMaxVelocity(vel);
					stepper.setSteps(steps);
				} catch(TinkerforgeException e) {
				}
			}
		});

		stepper.enable(); // Enable motor power
		stepper.setSteps(1); // Drive one step forward to get things going

		System.out.println("Press key to exit"); System.in.read();
		stepper.disable();
		ipcon.disconnect();
	}
}
